source/service.go

```diff
@@ -0,0 +1,63 @@
+package source
+
+import (
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api/v1"
+
+	"github.com/kubernetes-incubator/external-dns/endpoint"
+)
+
+const (
+	// The annotation used for figuring out which controller is responsible
+	controllerAnnotationKey = "external-dns.kubernetes.io/controller"
+	// The annotation used for defining the desired hostname
+	hostnameAnnotationKey = "external-dns.kubernetes.io/hostname"
+	// The value of the controller annotation so that we feel resposible
+	controllerAnnotationValue = "dns-controller"
+)
+
+// ServiceSource is an implementation of Source for Kubernetes service objects.
+// It will find all services that are under our jurisdiction, i.e. annotated
+// desired hostname and matching or no controller annotation. For each of the
+// matched services' external entrypoints it will return a corresponding
+// Endpoint object.
+type ServiceSource struct {
```

Comment:
Recently I used to add the suffix. Mostly inspired by other code, e.g. https://github.com/hashicorp/vault/blob/master/physical/s3.go#L24.